jupyter_client/consoleapp.py

```diff
@@ -311,6 +312,8 @@ def init_kernel_manager(self) -> None:
             self.exit(1)  # type:ignore[attr-defined]
 
         self.kernel_manager = t.cast(KernelManager, self.kernel_manager)
+
+        assert not iscoroutinefunction(self.kernel_manager.start_kernel)
```

Comment:
Yes I know asserts are removed in optimized mode, but they are useful in non-optimized mode to debug broken invariant. That's the point of assert.

I was hitting a "non awaited coroutine" in a weird configuration, this make the failure explicit.

I don't know where this "using assert it bad because they are removed in optimized build", I mean that's the point of asserts right ? From the [python docs](
https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement)

> Assert statements are a convenient way to insert debugging assertions into a program

Reply:
> But now you have code that behaves differently depending on a setting that you don't control.
> You can replace it with a check and raise an exception with a meaningful description of the error.

Yeah, but I don't want to have a permanent check for something that is never supposed to happen.

Reply:
But you have it for users who don't run in optimized mode.